In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('https://query.data.world/s/ux7gst37kd3ybbdq7mz74sbuw6uory?dws=00000')
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge
0,1.0,2020-01-01 00:28:15,2020-01-01 00:33:03,1.0,1.2,1.0,N,238,239,1.0,6.0,3.0,0.5,1.47,0.0,0.3,11.27,2.5
1,1.0,2020-01-01 00:35:39,2020-01-01 00:43:04,1.0,1.2,1.0,N,239,238,1.0,7.0,3.0,0.5,1.50,0.0,0.3,12.30,2.5
2,1.0,2020-01-01 00:47:41,2020-01-01 00:53:52,1.0,0.6,1.0,N,238,238,1.0,6.0,3.0,0.5,1.00,0.0,0.3,10.80,2.5
3,1.0,2020-01-01 00:55:23,2020-01-01 01:00:14,1.0,0.8,1.0,N,238,151,1.0,5.5,0.5,0.5,1.36,0.0,0.3,8.16,0.0
4,2.0,2020-01-01 00:01:58,2020-01-01 00:04:16,1.0,0.0,1.0,N,193,193,2.0,3.5,0.5,0.5,0.00,0.0,0.3,4.80,0.0


In [3]:
df.shape

(6405008, 18)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6405008 entries, 0 to 6405007
Data columns (total 18 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   VendorID               float64
 1   tpep_pickup_datetime   object 
 2   tpep_dropoff_datetime  object 
 3   passenger_count        float64
 4   trip_distance          float64
 5   RatecodeID             float64
 6   store_and_fwd_flag     object 
 7   PULocationID           int64  
 8   DOLocationID           int64  
 9   payment_type           float64
 10  fare_amount            float64
 11  extra                  float64
 12  mta_tax                float64
 13  tip_amount             float64
 14  tolls_amount           float64
 15  improvement_surcharge  float64
 16  total_amount           float64
 17  congestion_surcharge   float64
dtypes: float64(13), int64(2), object(3)
memory usage: 879.6+ MB


In [5]:
df.describe()

,VendorID,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge
count,6.339567e+06,6.339567e+06,6.405008e+06,6.339567e+06,6.405008e+06,6.405008e+06,6.339567e+06,6.405008e+06,6.405008e+06,6.405008e+06,6.405008e+06,6.405008e+06,6.405008e+06,6.405008e+06,6.405008e+06
mean,1.669624e+00,1.515333e+00,2.929644e+00,1.059908e+00,1.647323e+02,1.626627e+02,1.270298e+00,1.269411e+01,1.115456e+00,4.923182e-01,2.189342e+00,3.488395e-01,2.979870e-01,1.866315e+01,2.275662e+00
std,4.703484e-01,1.151594e+00,8.315911e+01,8.118432e-01,6.554374e+01,6.991261e+01,4.739985e-01,1.212730e+01,1.260054e+00,7.374184e-02,2.760028e+00,1.766978e+00,3.385937e-02,1.475736e+01,7.352646e-01
min,1.000000e+00,0.000000e+00,-3.062000e+01,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,-1.238000e+03,-2.700000e+01,-5.000000e-01,-9.100000e+01,-3.574000e+01,-3.000000e-01,-1.242300e+03,-2.500000e+00
25%,1.000000e+00,1.000000e+00,9.600000e-01,1.000000e+00,1.320000e+02,1.130000e+02,1.000000e+00,6.500000e+00,0.000000e+00,5.000000e-01,0.000000e+00,0.000000e+00,3.000000e-01,1.116000e+01,2.500000e+00
50%,2.000000e+00,1.000000e+00,1.600000e+00,1.000000e+00,1.620000e+02,1.620000e+02,1.000000e+00,9.000000e+00,5.000000e-01,5.000000e-01,1.950000e+00,0.000000e+00,3.000000e-01,1.430000e+01,2.500000e+00
75%,2.000000e+00,2.000000e+00,2.930000e+00,1.000000e+00,2.340000e+02,2.340000e+02,2.000000e+00,1.400000e+01,2.500000e+00,5.000000e-01,2.860000e+00,0.000000e+00,3.000000e-01,1.980000e+01,2.500000e+00
max,2.000000e+00,9.000000e+00,2.102401e+05,9.900000e+01,2.650000e+02,2.650000e+02,5.000000e+00,4.265000e+03,1.130100e+02,3.080000e+01,1.100000e+03,9.105000e+02,3.000000e-01,4.268300e+03,2.750000e+00


In [6]:
df.isnull().sum()

,0
VendorID,65441
tpep_pickup_datetime,0
tpep_dropoff_datetime,0
passenger_count,65441
trip_distance,0
RatecodeID,65441
store_and_fwd_flag,65441
PULocationID,0
DOLocationID,0
payment_type,65441


In [7]:
(65441/len(df))*100

1.021716132126611

There is only 1.0217% of data is null in dataset. So we can directly remove the data from Dataset. It will not impact any predication,


(No of null_values/len(df))*100


This formula helps determine the percentage of **null** (missing) values present in the dataset

In [8]:
df = df[df['fare_amount'] >= 0]

This will make sure that 'fare_amount' is not a negative value

In [9]:
df["payment_type"].unique()

array([ 1.,  2.,  4.,  3.,  5., nan])

In [10]:
df.isna().sum()

,0
VendorID,65377
tpep_pickup_datetime,0
tpep_dropoff_datetime,0
passenger_count,65377
trip_distance,0
RatecodeID,65377
store_and_fwd_flag,65377
PULocationID,0
DOLocationID,0
payment_type,65377


The features such as `store_and_fwd_flag` , `RatecodeID` , `payment_type` contain a large amount of NaN values.<br>
So we ned to check is there any relationship between these features and the target feature `total_amount`<br>
If not, we can remove them

Checking if the features have a string relationship with the target variable

In [11]:
features = ['store_and_fwd_flag', 'RatecodeID', 'payment_type']

for feature in features:
    print(df[feature].unique())

['N' 'Y' nan]
[ 1.  5.  3.  2.  4. 99.  6. nan]
[ 1.  2.  4.  3.  5. nan]


In [12]:
from scipy import stats
from scipy.stats import f_oneway
from scipy.stats import zscore

data = df[['store_and_fwd_flag','RatecodeID','payment_type','total_amount']].dropna()

# store_and_fwd_flag
group_n = data[data['store_and_fwd_flag'] == 'N']['total_amount']
group_y = data[data['store_and_fwd_flag'] == 'Y']['total_amount']

# RatecodeID
rate_groups = [
    data[data['RatecodeID']==i]['total_amount']
    for i in sorted(data['RatecodeID'].unique())
]

# payment_type
group_p1 = data[data['payment_type'] == 1]['total_amount']
group_p2 = data[data['payment_type'] == 2]['total_amount']

# t-testss
t1,p1 = stats.ttest_ind(group_n,group_y,equal_var=False)
t2,p2 = stats.ttest_ind(group_p1,group_p2,equal_var=False)

#Effect_Size
effect_size = (group_n.mean() - group_y.mean()) / data['total_amount'].std()
print("Effect size(store_and_fwd_flag):", effect_size)
effect_size_pt = (group_p1.mean() - group_p2.mean()) / data['total_amount'].std()
print("Effect size (payment_type):", effect_size_pt)

#Anova test(f-test)
f1,p3 = f_oneway(*rate_groups)

print(f"T1:{t1:.4f} P1:{p1:.4f}")
print(f"T2:{t2:.4f} P2:{p2:.4f}")
print(f"F1:{f1:.4f} P3:{p3:.4f}")

#null hypothesis
null_Hypothesis_SFF = "There is no difference in total_amount between store_and_fwd_flag 'N' and 'Y'"
null_Hypothesis_PT = "There is no difference in total_amount between payment_type '1' and '2'"

if p1 < 0.05:
    print(f"Reject null hypothesis: {null_Hypothesis_SFF}")
else:
    print(f"Fail to reject null hypothesis {null_Hypothesis_SFF}")

if p2 < 0.05:
    print(f"Reject null hypothesis: {null_Hypothesis_PT}")
else:
    print(f"Fail to reject null hypothesis {null_Hypothesis_PT}")


alpha = 0.05
if p3 < alpha:
    print("Reject null hypothesis: There is a relationship between RatecodeID and total_amount")
else:
    print("Fail to reject null hypothesis: There might not be a relationship between RatecodeID and total_amount")


Effect size(store_and_fwd_flag): 0.02455461918137002
Effect size (payment_type): 0.2762960404760413
T1:6.7497 P1:0.0000
T2:330.7180 P2:0.0000
F1:896801.7024 P3:0.0000
Reject null hypothesis: There is no difference in total_amount between store_and_fwd_flag 'N' and 'Y'
Reject null hypothesis: There is no difference in total_amount between payment_type '1' and '2'
Reject null hypothesis: There is a relationship between RatecodeID and total_amount


For the feature `store_and_fwd_flag` and `payment_type`. t-test is used to determine if there is any relation between them and `total_amount`.<br>

For the feature 'RatecodeID'.Anova test(f-test) is performed to determind the relation between itself and `total_amount`.<br>

The P-value is 0 coz of the Large dataset.
Thats why effect_size shows how strong or meaning full the effect is

Effect size measures the magnitude of difference between groups.

Droping NaN values from the features


In [13]:
df.dropna(inplace=True)
df.reset_index(drop=True, inplace=True)
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge
0,1.0,2020-01-01 00:28:15,2020-01-01 00:33:03,1.0,1.2,1.0,N,238,239,1.0,6.0,3.0,0.5,1.47,0.0,0.3,11.27,2.5
1,1.0,2020-01-01 00:35:39,2020-01-01 00:43:04,1.0,1.2,1.0,N,239,238,1.0,7.0,3.0,0.5,1.50,0.0,0.3,12.30,2.5
2,1.0,2020-01-01 00:47:41,2020-01-01 00:53:52,1.0,0.6,1.0,N,238,238,1.0,6.0,3.0,0.5,1.00,0.0,0.3,10.80,2.5
3,1.0,2020-01-01 00:55:23,2020-01-01 01:00:14,1.0,0.8,1.0,N,238,151,1.0,5.5,0.5,0.5,1.36,0.0,0.3,8.16,0.0
4,2.0,2020-01-01 00:01:58,2020-01-01 00:04:16,1.0,0.0,1.0,N,193,193,2.0,3.5,0.5,0.5,0.00,0.0,0.3,4.80,0.0


In [14]:
df.isna().sum()

,0
VendorID,0
tpep_pickup_datetime,0
tpep_dropoff_datetime,0
passenger_count,0
trip_distance,0
RatecodeID,0
store_and_fwd_flag,0
PULocationID,0
DOLocationID,0
payment_type,0


# Feature Engineering

In [15]:
#Converting the pickup and dropoff datetime columns to datetime type
df['tpep_pickup_datetime'] = pd.to_datetime(df['tpep_pickup_datetime'])
df['tpep_dropoff_datetime'] = pd.to_datetime(df['tpep_dropoff_datetime'])

In [16]:
#Converting the dtype of the features
df['fare_amount'] = df['fare_amount'].astype('int64')
df['payment_type'] = df['payment_type'].astype('int64')

In [17]:
#Adding columns of pickup
df['pickup_hours'] = df['tpep_pickup_datetime'].dt.hour
df['pickup_day'] = df['tpep_pickup_datetime'].dt.day
df['pickup_weekday'] = df['tpep_pickup_datetime'].dt.weekday
df['pickup_month'] = df['tpep_pickup_datetime'].dt.month

It important to know the `pickup_hourse` , `pickup_day` , `pickup_weekday` , `pickupmonth` <br> Taxi Fare depend on the time patterns

In [18]:
# Calculating trip duration
df['trip_duration(min)'] = ((df['tpep_dropoff_datetime'] - df['tpep_pickup_datetime']).dt.total_seconds()/60).astype(int)
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,pickup_hours,pickup_day,pickup_weekday,pickup_month,trip_duration(min)
0,1.0,2020-01-01 00:28:15,2020-01-01 00:33:03,1.0,1.2,1.0,N,238,239,1,...,1.47,0.0,0.3,11.27,2.5,0,1,2,1,4
1,1.0,2020-01-01 00:35:39,2020-01-01 00:43:04,1.0,1.2,1.0,N,239,238,1,...,1.50,0.0,0.3,12.30,2.5,0,1,2,1,7
2,1.0,2020-01-01 00:47:41,2020-01-01 00:53:52,1.0,0.6,1.0,N,238,238,1,...,1.00,0.0,0.3,10.80,2.5,0,1,2,1,6
3,1.0,2020-01-01 00:55:23,2020-01-01 01:00:14,1.0,0.8,1.0,N,238,151,1,...,1.36,0.0,0.3,8.16,0.0,0,1,2,1,4
4,2.0,2020-01-01 00:01:58,2020-01-01 00:04:16,1.0,0.0,1.0,N,193,193,2,...,0.00,0.0,0.3,4.80,0.0,0,1,2,1,2


## Reordering and Renaming Columns

In [19]:
df = df[[
    'VendorID','passenger_count','trip_distance','RatecodeID','store_and_fwd_flag',
    'PULocationID','DOLocationID','payment_type','fare_amount','extra','mta_tax',
    'tip_amount','tolls_amount','improvement_surcharge','total_amount',
    'congestion_surcharge','trip_duration(min)'
    ]]

df.columns = [
    'vendor','passengers','distance','rate_id','flag','pickup_id','dropoff_id',
    'payment_type','fare','extras','tax','tip','tolls','improvement','total',
    'congestion','duration'
    ]

df['flag'] = df['flag'].map({'N':0,'Y':1})
df.head()

,vendor,passengers,distance,rate_id,flag,pickup_id,dropoff_id,payment_type,fare,extras,tax,tip,tolls,improvement,total,congestion,duration
0,1.0,1.0,1.2,1.0,0,238,239,1,6,3.0,0.5,1.47,0.0,0.3,11.27,2.5,4
1,1.0,1.0,1.2,1.0,0,239,238,1,7,3.0,0.5,1.50,0.0,0.3,12.30,2.5,7
2,1.0,1.0,0.6,1.0,0,238,238,1,6,3.0,0.5,1.00,0.0,0.3,10.80,2.5,6
3,1.0,1.0,0.8,1.0,0,238,151,1,5,0.5,0.5,1.36,0.0,0.3,8.16,0.0,4
4,2.0,1.0,0.0,1.0,0,193,193,2,3,0.5,0.5,0.00,0.0,0.3,4.80,0.0,2


## Removing Duplicate values

In [20]:
print(df.duplicated().sum())

565438


In [21]:
#dropping the duplicates and reseting the index
df.drop_duplicates(inplace=True)
df.reset_index(drop=True, inplace=True)

#Cheching again after dropping the duplicates
print(df.duplicated().sum())

0


In [22]:
df.dtypes

,0
vendor,float64
passengers,float64
distance,float64
rate_id,float64
flag,int64
pickup_id,int64
dropoff_id,int64
payment_type,int64
fare,int64
extras,float64


### Specifying numerical and categorical columns


In [23]:
cat_cols = [
    'vendor','rate_id','flag',
    'pickup_id','dropoff_id','payment_type'
    ]

num_cols = [
    'passengers','distance','fare','extras','tax','tip','tolls',
    'improvement','congestion','duration'
    ]

### Standerdizing Data

In [24]:
from sklearn.preprocessing import StandardScaler

#Standerd Scaling of numerical columns
scaler = StandardScaler()
df[num_cols] = scaler.fit_transform(df[num_cols])

df.head()

,vendor,passengers,distance,rate_id,flag,pickup_id,dropoff_id,payment_type,fare,extras,tax,tip,tolls,improvement,total,congestion,duration
0,1.0,-0.46783,-0.471622,1.0,0,238,239,1,-0.566021,1.56992,0.076877,-0.290247,-0.200762,0.024506,11.27,0.290215,-0.184364
1,1.0,-0.46783,-0.471622,1.0,0,239,238,1,-0.483679,1.56992,0.076877,-0.279720,-0.200762,0.024506,12.30,0.290215,-0.139239
2,1.0,-0.46783,-0.626382,1.0,0,238,238,1,-0.566021,1.56992,0.076877,-0.455177,-0.200762,0.024506,10.80,0.290215,-0.154281
3,1.0,-0.46783,-0.574796,1.0,0,238,151,1,-0.648364,-0.45274,0.076877,-0.328848,-0.200762,0.024506,8.16,-3.445780,-0.184364
4,2.0,-0.46783,-0.781143,1.0,0,193,193,2,-0.813048,-0.45274,0.076877,-0.806092,-0.200762,0.024506,4.80,-3.445780,-0.214448


### Splitting Data

In [25]:
from sklearn.model_selection import train_test_split

#Spliting data into feature and target
X = df[cat_cols + num_cols]
y = df['total']

#Split data into Traning and Test Sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, shuffle = False )

### Deciison Tree Regressor

In [26]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Train the DecisionTreeRegressor model
dt_Model = DecisionTreeRegressor(random_state=42)
dt_Model.fit(X_train, y_train)

#Predict on the test set
y_pred = dt_Model.predict(X_test)

#Evaluate the model
mae_dt = mean_absolute_error(y_test, y_pred)
mse_dt = mean_squared_error(y_test, y_pred)
rmse_dt = np.sqrt(mse_dt)
r2_dt = r2_score(y_test, y_pred)

#Printing the Values
print(f'Mean Absolute Error: {mae_dt}')
print(f'Mean Squared Error: {mse_dt}')
print(f'Root Mean Squared Error: {rmse_dt}')
print(f'R-squared: {r2_dt}')

Mean Absolute Error: 0.09169191708577246
Mean Squared Error: 0.7343755284071676
Root Mean Squared Error: 0.8569571333545031
R-squared: 0.9965764283968577


### Random Forest Regressor

In [27]:
from sklearn.ensemble import RandomForestRegressor

#Train the RandomForestRegressor model
rfr_Model = RandomForestRegressor(
    n_estimators=20,
    max_depth=10,
    n_jobs=-1,
    random_state=42
)
rfr_Model.fit(X_train, y_train)

#Predict on the test set
y_pred = rfr_Model.predict(X_test)

#Evaluate the test set
mae_rfr = mean_absolute_error(y_test, y_pred)
mse_rfr = mean_squared_error(y_test, y_pred)
rmse_rfr = np.sqrt(mse_rfr)
r2_rfr = r2_score(y_test, y_pred)

#Printing the Values
print(f'Mean Absolute Error: {mae_rfr}')
print(f'Mean Squared Error: {mse_rfr}')
print(f'Root Mean Squaared Error: {rmse_rfr}')
print(f'R-squared: {r2_rfr}')

Mean Absolute Error: 0.3914764913104171
Mean Squared Error: 1.351639518815868
Root Mean Squaared Error: 1.1626003263442979
R-squared: 0.9936988168923061


### Gradient Boosting Regressor

In [28]:
from sklearn.ensemble import GradientBoostingRegressor

#Train the GradientBootingRegressor model
gbr_Model = GradientBoostingRegressor(
    n_estimators=50,
    max_depth=5,
    learning_rate=0.1,
    random_state=42
)
gbr_Model.fit(X_train, y_train)

#Predict on the test set
y_pred_gbr = gbr_Model.predict(X_test)

#Evaluating the test set
mae_gbr = mean_absolute_error(y_test, y_pred_gbr)
mse_gbr = mean_squared_error(y_test, y_pred_gbr)
rmse_gbr = np.sqrt(mse_gbr)
r2_gbr = r2_score(y_test, y_pred_gbr)

#Printing the Values
print(f'Mean Absolute Error: {mae_gbr}')
print(f'Mean Squared Error: {mse_gbr}')
print(f'Root Mean Squaared Error: {rmse_gbr}')
print(f'R-squared: {r2_gbr}')

Mean Absolute Error: 0.35314929752074503
Mean Squared Error: 0.7423018501261239
Root Mean Squaared Error: 0.8615694110900897
R-squared: 0.9965394768252097


In [29]:
import xgboost as xgb

# Train the XGBoost Regressor model
xgb_model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)
xgb_model.fit(X_train, y_train)

# Predict on the test set
y_pred_xgb = xgb_model.predict(X_test)

# Evaluate the model
mae_xgb = mean_absolute_error(y_test, y_pred_xgb)
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
rmse_xgb = np.sqrt(mse_xgb)
r2_xgb = r2_score(y_test, y_pred_xgb)

print("XGBoost Regressor Results:")
print(f'Mean Absolute Error: {mae_xgb}')
print(f'Mean Squared Error: {mse_xgb}')
print(f'Root Mean Squared Error: {rmse_xgb}')
print(f'R-squared: {r2_xgb}')

XGBoost Regressor Results:
Mean Absolute Error: 0.3261109872595791
Mean Squared Error: 8.377387735077674
Root Mean Squared Error: 2.8943717340862896
R-squared: 0.960945612089591


In [30]:
import numpy as np

random_number = np.random.randint(0, len(X_test))

test = X_test.iloc[[random_number]]

# printing predicted fares
print(f'Fare predicted with Decision Tree Regressor: {dt_Model.predict(test)[0]}')
print(f'Fare predicted with Random Forest Regressor: {rfr_Model.predict(test)[0]}')
print(f'Fare predicted with Gradient Boost Regressor: {gbr_Model.predict(test)[0]}')
print(f'Fare predicted with XGBoost Regressor: {xgb_model.predict(test)[0]}')

print('--------------------------------------------------------------')

print(f'The actual fare is: {y_test.iloc[random_number]}')

Fare predicted with Decision Tree Regressor: 15.360000000000436
Fare predicted with Random Forest Regressor: 15.125746539157172
Fare predicted with Gradient Boost Regressor: 15.230543930882233
Fare predicted with XGBoost Regressor: 15.193415641784668
--------------------------------------------------------------
The actual fare is: 15.36
